docs: add "How to migrate from pytest-operator to Jubilant" (#98)
canonical/jubilant: docs/how-to/migrate-from-pytest-operator.md

# How to migrate from pytest-operator to Jubilant

Many charm integration tests use [pytest-operator](https://github.com/charmed-kubernetes/pytest-operator) and [python-libjuju](https://github.com/juju/python-libjuju). This guide explains how to migrate your integration tests from those libraries to Jubilant.

To get help while you're migrating tests, please keep the [API reference](/reference/jubilant) handy, and make use of your IDE's autocompletion -- Jubilant tries to provide good type annotations and docstrings.

Migrating your tests can be broken into three steps:

1. Update your dependencies
2. Add fixtures to `conftest.py`
3. Update the tests themselves

Let's look at each of these in turn.

## Update your dependencies

> The first thing you'll need to do is add `jubilant` as a dependency to your `tox.ini` or `pyproject.toml` dependencies. Before Jubilant 1.0 is released, you should consider locking it to a specific version.

Reviewer: How about strengthening "you should consider" to "we recommend"?

Author: Agreed, done.

> You can also remove the dependencies on `juju` (python-libjuju), `pytest-operator`, and `pytest-asyncio`.

Author: I think it's probably better to show the diff all together, but I've added a paragraph after the diff on this:

> If you're migrating a large number of tests, you may want to do it in stages. In that case, keep the old dependencies in place till the end, and migrate tests one at a time, so that both pytest-operator and Jubilant tests can run together.

You can optionally provide an *error* callable, which also takes a `Status`, and if that returns True, `wait` raises a [`WaitError`](jubilant.WaitError) immediately.

Jubilant provides helper functions to use for the *ready* and *error* callables, such as [`jubilant.all_active`](jubilant.all_active) and [`jubilant.any_error`](jubilant.any_error). These check whether the status of all (or any) applications and their units are in a given state.

For example, a simple `wait` call that waits for all applications and units to go "active" and raises an error if any go into "error", would look like this:

```python
# pytest-operator
async def test_active(model: Model):
    await model.deploy('mycharm')
    await model.wait_for_idle(status='active')  # implies raise_on_error=True

# jubilant
def test_active(juju: jubilant.Juju):
    juju.deploy('mycharm')
    juju.wait(jubilant.all_active, error=jubilant.any_error)
```

It's common to use a `lambda` function to customize the callables further. For example, to wait specifically till `mysql` and `redis` are active:

```python
juju.wait(lambda status: jubilant.all_active(status, ['mysql', 'redis']))
```

> The `wait` method also has other options (see the [reference docs](jubilant.Juju.wait)):

Reviewer: ```suggestion
The `wait` method also has other options (see [`juju.wait`](jubilant.Juju.wait) for details):
```

```python
juju.deploy('mycharm')
juju.wait(
    ready=lambda status: status.apps['mycharm'].is_active,
    error=jubilant.any_error,
    delay=0.2,    # poll "juju status" every 200ms (default 1s)
    timeout=60,   # set overall timeout to 60s (default juju.wait_timeout)
    successes=7,  # require ready to return success 7x in a row (default 3)
)
```

### Integrating two applications

To integrate two charms, remove the `async`-related code and replace `model.add_relation` with [`juju.integrate`](jubilant.Juju.integrate). For example, to integrate discourse-k8s with three other charms:

```python
# pytest-operator
await asyncio.gather(
    model.add_relation('discourse-k8s', 'postgresql-k8s:database'),
    model.add_relation('discourse-k8s', 'redis-k8s'),
    model.add_relation('discourse-k8s', 'nginx-ingress-integrator'),
)
await model.wait_for_idle(status='active')

# jubilant
juju.integrate('discourse-k8s', 'postgresql-k8s:database')
juju.integrate('discourse-k8s', 'redis-k8s')
juju.integrate('discourse-k8s', 'nginx-ingress-integrator')
juju.wait(jubilant.all_active)
```

### Executing a command

> In `pytest-operator` tests, you used `unit.run` to execute a command. With Jubilant (like Juju 3.x) you use [`juju.exec`](jubilant.Juju.exec). Jubilant's `exec` returns a [`jubilant.Task`](jubilant.Task), and it also checks errors for you:

Reviewer: ```suggestion
In `pytest-operator` tests, you used `unit.run` to execute a command. With Jubilant (as with Juju 3.x) you use [`juju.exec`](jubilant.Juju.exec). Jubilant's `exec` returns a [`jubilant.Task`](jubilant.Task), and it also checks errors for you:
```

```python
# pytest-operator
unit = model.applications['discourse-k8s'].units[0]
action = await unit.run('/bin/bash -c "..."')
await action.wait()
logger.info(action.results)
assert action.results['return-code'] == 0, 'Enable plugins failed'

# jubilant
task = juju.exec('/bin/bash -c "..."', unit='discourse-k8s/0')
logger.info(task.results)
```

### Running an action

In `pytest-operator` tests, you used `unit.run_action` to run an action. With Jubilant, you use [`juju.run`](jubilant.Juju.run). Similar to `exec`, Jubilant's `run` returns a [`jubilant.Task`](jubilant.Task) and checks errors for you:

```python
# pytest-operator
app = model.applications['postgresl-k8s']
action = await app.units[0].run_action('get-password', username='operator')
await action.wait()
password = action.results['password']

# jubilant
task = juju.run('postgresql-k8s/0', 'get-password', {'username': 'operator'})
password = task.results['password']
```

### The `cli` fallback

> Similar to how you could call `ops_test.juju`, with Jubilant you can call [`juju.cli`](jubilant.Juju.cli) as a fallback to execute an arbitrary Juju command. Once again, it checks errors for you (and raises [`CLIError`](jubilant.CLIError) if the command's exit code is nonzero):

Reviewer: ```suggestion
Similar to how you could call `ops_test.juju`, with Jubilant you can call [`juju.cli`](jubilant.Juju.cli) to execute an arbitrary Juju command. The `cli` method checks errors for you and raises a [`CLIError`](jubilant.CLIError) if the command's exit code is nonzero:
```

```python
# pytest-operator
return_code, _, scp_err = await ops_test.juju(
    'scp',
    '--container',
    'postgresql',
    './testing_database/testing_database.sql',
    f'{postgres_app.units[0].name}:.',
)
assert return_code == 0, scp_err

# jubilant
juju.cli(
    'scp',
    '--container',
    'postgresql',
    './testing_database/testing_database.sql',
    'postgresql-k8s/0:.',
)
```

> ## Further information

Reviewer: ```suggestion
## See more
```

> - [Jubilant's API reference](jubilant)

Reviewer: ```suggestion
- [Jubilant's API reference](/reference/jubilant)
```

I think it's better to use a page link rather than linking to the top-level of the auto-generated reference. They both work, but without any `.` it's less predictable what this is going to link to.

> - [This discourse-k8s migration PR](https://github.com/canonical/discourse-k8s-operator/pull/326) shows what migrating a real charm's integration tests looks like

Reviewer: ```suggestion
- [This discourse-k8s migration PR](https://github.com/canonical/discourse-k8s-operator/pull/326) shows how we migrated a real charm's integration tests
```
